### PR TITLE
[wasm][debugger] Fix reading metadata info after applying changes

### PR DESCRIPTION
**Pull request by thaystg**
Doing the same thing that is done for GetString from metadatareader for:
GetTypeDefinition
GetTypeReference
GetMethodDefinition
GetMemberReference

It was also needed another change because methodDef.GetCustomAttributes() does not work on EnC metadatas.

The parent rowid used on CustomAttributes is the accumulated one(methodId from assembly + methodId from EnC), while the one used to get the MethodDef is the EnC rowid. 
So when we do this: methodDef.GetCustomAttributes() it doesn't find any CustomAttribute as it's using rowId 3 for example, and it should use rowId 19 which is the accumulated one.

### CODE REVIEW

**Review by lewing — APPROVED** [review]

File: src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs [hunk 1]
```diff
@@ -1028,6 +1027,61 @@ private static int GetMethodDebugInformationIdx(MetadataReader pdbMetadataReader
             return -1;
         }
 
+        public TypeReference EnCGetTypeReference(TypeReferenceHandle methodDefHandle)
+        {
+            var asmMetadataReaderLocal = asmMetadataReader;
```
Comment (radical) [inline src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:1032]: Curious, why is this needed?

File: src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs [hunk 2]
```diff
@@ -949,16 +948,16 @@ public bool TryGetCustomAttributeName(CustomAttributeHandle customAttribute, Met
...
                     _ => null,
                 };
                 if (container == null)
                     return false;
                 StringHandle? attributeTypeNameHandle = container.Value.Kind switch
                 {
-                    HandleKind.TypeDefinition => metadataReader.GetTypeDefinition((TypeDefinitionHandle)container.Value).Name,
-                    HandleKind.TypeReference => metadataReader.GetTypeReference((TypeReferenceHandle)container.Value).Name,
+                    HandleKind.TypeDefinition => EnCGetTypeDefinition((TypeDefinitionHandle)container.Value).Name,
+                    HandleKind.TypeReference => EnCGetTypeReference((TypeReferenceHandle)container.Value).Name,
```
Comment (radical) [inline src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:960]: - Do IIUC that these new methods would work irrespective of whether EnC was used with the assembly or not?
- Should we check the other places where we currently use `*.GetTypeDefinition` etc, to see if they need to use these new methods too?

File: src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs (hunk 1 above)
Comment (radical) [reply src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:1032]: nvm.. I see it at line 1038!! 👍

File: src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs (hunk 2 above)
Comment (thaystg) [reply src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:964]: 1) Yes
2) I checked and I think we should use only in this place.

File: src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs [hunk 3]
```diff
@@ -389,11 +389,14 @@ public MethodInfo(AssemblyInfo assembly, MethodDefinitionHandle methodDefHandle,
             this.IsEnCMethod = false;
             this.TypeInfo = type;
             DebuggerAttrInfo = new DebuggerAttributesInfo();
-            foreach (CustomAttributeHandle cattr in methodDef.GetCustomAttributes())
+            //we need to loop in all the CustomAttributes from asmMetadataReader because methodDef.GetCustomAttributes() does not work correctly on EnC metadata
+            foreach (CustomAttributeHandle cattr in asmMetadataReader.CustomAttributes)
             {
+                var ca = asmMetadataReader.GetCustomAttribute(cattr);
+                if (ca.Parent.Kind != HandleKind.MethodDefinition || (ca.Parent.Kind == HandleKind.MethodDefinition && ca.Parent.GetHashCode() != (token | (int)TokenType.MdtMethodDef)))
```
Comment (radical) [inline src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs]: Could you please add a comment here explaining what this is trying to do?

File: src/mono/wasm/debugger/tests/ApplyUpdateReferencedAssembly/MethodBody1.cs [hunk 4]
```diff
@@ -77,6 +77,24 @@ public class MethodBody6 {
     
     
     
+
```
Comment (radical) [inline src/mono/wasm/debugger/tests/ApplyUpdateReferencedAssembly/MethodBody1.cs:80]: Can we replace the empty lines with an empty comment block saying that this shouldn't be changed?

File: src/mono/wasm/debugger/DebuggerTestSuite/HotReloadTests.cs [hunk 5]
```diff
@@ -600,6 +600,32 @@ public async Task DebugHotReload_NewInstanceFields()
...
+        [ConditionalFact(nameof(RunningOnChrome))]
+        public async Task DebugHotReloadMethod_AddingNewClassUsingDebugAttribute()
+        {
+            await SetJustMyCode(true);
+            string asm_file = Path.Combine(DebuggerTestAppPath, "ApplyUpdateReferencedAssembly.dll");
+            string pdb_file = Path.Combine(DebuggerTestAppPath, "ApplyUpdateReferencedAssembly.pdb");
+            string asm_file_hot_reload = Path.Combine(DebuggerTestAppPath, "../wasm/ApplyUpdateReferencedAssembly.dll");
+
+            var bp_notchanged = await SetBreakpoint(".*/MethodBody1.cs$", 55, 12, use_regex: true);
+            var bp_invalid = await SetBreakpoint(".*/MethodBody1.cs$", 118, 12, use_regex: true);
```
Comment (radical) [inline src/mono/wasm/debugger/DebuggerTestSuite/HotReloadTests.cs:612]: unused

File: src/mono/wasm/debugger/DebuggerTestSuite/HotReloadTests.cs [hunk 6]
```diff
@@ -600,6 +600,32 @@ public async Task DebugHotReload_NewInstanceFields()
...
 
+        [ConditionalFact(nameof(RunningOnChrome))]
+        public async Task DebugHotReloadMethod_AddingNewClassUsingDebugAttribute()
+        {
+            await SetJustMyCode(true);
+            string asm_file = Path.Combine(DebuggerTestAppPath, "ApplyUpdateReferencedAssembly.dll");
+            string pdb_file = Path.Combine(DebuggerTestAppPath, "ApplyUpdateReferencedAssembly.pdb");
+            string asm_file_hot_reload = Path.Combine(DebuggerTestAppPath, "../wasm/ApplyUpdateReferencedAssembly.dll");
+
+            var bp_notchanged = await SetBreakpoint(".*/MethodBody1.cs$", 55, 12, use_regex: true);
```
[the remark this answers is not in the document]
Comment (radical) [reply src/mono/wasm/debugger/DebuggerTestSuite/HotReloadTests.cs:611]: I mean the *local*

File: src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs (hunk 3 above)
Comment (radical) [reply src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs]: `if (fromEnC && (ca.Parent.Kind != HandleKind.MethodDefinition || ca.Parent.GetHashCode() != (token | (int)TokenType.MdtMethodDef)))`

File: src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs [hunk 7]
```diff
@@ -845,11 +849,11 @@ internal sealed class AssemblyInfo
         internal MetadataReader asmMetadataReader { get; }
         internal MetadataReader pdbMetadataReader { get; set; }
 
-        internal List<Tuple<MetadataReader, MetadataReader>> enCMetadataReader  = new List<Tuple<MetadataReader, MetadataReader>>();
+        internal List<Tuple<MetadataReader, MetadataReader>> enCMetadataReader = new List<Tuple<MetadataReader, MetadataReader>>();
```
Comment (radical) [inline src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:852]:
```suggestion
        internal List<Tuple<MetadataReader, MetadataReader>> enCMetadataReader = new();
```

File: src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs [hunk 8]
```diff
@@ -845,11 +849,11 @@ internal sealed class AssemblyInfo
         internal MetadataReader asmMetadataReader { get; }
         internal MetadataReader pdbMetadataReader { get; set; }
 
-        internal List<Tuple<MetadataReader, MetadataReader>> enCMetadataReader  = new List<Tuple<MetadataReader, MetadataReader>>();
+        internal List<Tuple<MetadataReader, MetadataReader>> enCMetadataReader = new();
```
Comment (radical) [inline src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:852]: Sorry, one more:

```suggestion
        internal List<(MetadataReader asm, MetadataReader pdb)> enCMetadataReader = new();
```

This will ensure that we don't mistakenly get the wrong member.

File: src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs [hunk 9]
```diff
@@ -370,7 +370,7 @@ public MethodInfo(AssemblyInfo assembly, string methodName, int methodToken, Typ
             _asyncScopes = Array.Empty<AsyncScopeDebugInformation>();
         }
 
-        public MethodInfo(AssemblyInfo assembly, MethodDefinitionHandle methodDefHandle, int token, SourceFile source, TypeInfo type, MetadataReader asmMetadataReader, MetadataReader pdbMetadataReader)
+        public MethodInfo(AssemblyInfo assembly, MethodDefinitionHandle methodDefHandle, int token, SourceFile source, TypeInfo type, MetadataReader asmMetadataReader, MetadataReader pdbMetadataReader, bool fromEnC = false)
```
Comment (radical) [inline src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:373]: nit: remove the default value for `fromEnC`. This would mean that any code calling this constructor would have to decide whether it is enc or not, which should be useful for new code.

**Review by radical — APPROVED** [review]
LGTM 👍
